Feature / BingX connector

danilo-robotter-ai opened the pull request:
We are adding a new connector for the BingX exchange.

The manual and automated tests are passing:
 - adding a new wallet
 - checking balances
 - starting and using a PMM v1 and v2 strategy
 - unit tests

nikspz commented:
https://snapshot.org/#/hbot-ncp.eth/proposal/0xac46cb41a8f66e7cb13f99471da7c8d1bd2c47f81e3743be60e667638e4ac436
Approved

nikspz commented:
6fe044c
Test performed:
- Cloned and installed PR7280 
- connected bing_x API keys successfully ✅
- checked balance successfully, matched with exchange ✅
- created/started pureMM and pmm_simple strategies: ok ✅
- Review orders price/ amount could be inconsistent (4.99 instead of 5, 0.99 instead of 1 etc.) ❌
- Failed to cancel events during the strategy run (order stuck on exchange❌
- After stop command always returns `Failed to cancel the following orders on bing_x` (for PureMM v1 strategy) ❌

Steps:

1.  Cloned and installed PR7280
2. connect bing_x
3. create/start pureMM strategy using bing_x

Actual:

- Review orders price/ amount could be inconsistent (4.99 instead of 5, 0.99 instead of 1 etc.)
- After stop command always returns `Failed to cancel the following orders on bing_x`


![bingx-inconsistent-order_amount_and-price](https://github.com/user-attachments/assets/bbe21ed4-3bc5-4451-b507-762e58fd42bc)

![bingxstop-command-failed-to-cancel](https://github.com/user-attachments/assets/f54b878b-db55-46be-85d1-a281e2292773)
![image](https://github.com/user-attachments/assets/395b2c3c-f93a-47e7-afaf-545fc38759f4)

Failed to cancel event during strategy run:

[logs_conf_pure_mm_1.log](https://github.com/user-attachments/files/17924998/logs_conf_pure_mm_1.log)
[conf_pure_mm_1.zip](https://github.com/user-attachments/files/17925003/conf_pure_mm_1.zip)
[7280.zip](https://github.com/user-attachments/files/17925010/7280.zip)

```
2024-11-26 18:38:28,309 - 1960469 - hummingbot.connector.exchange.bing_x.bing_x_exchange.BingXExchange - ERROR - Failed to cancel order BINGX_SXPUT627d526849076f9311c15
Traceback (most recent call last):
  File "/home/hbot02/nikita/bingx2/hummingbot/connector/exchange_py_base.py", line 520, in _execute_order_cancel
    cancelled = await self._execute_order_cancel_and_process_update(order=order)
  File "/home/hbot02/nikita/bingx2/hummingbot/connector/exchange_py_base.py", line 540, in _execute_order_cancel_and_process_update
    cancelled = await self._place_cancel(order.client_order_id, order)
  File "/home/hbot02/nikita/bingx2/hummingbot/connector/exchange/bing_x/bing_x_exchange.py", line 229, in _place_cancel
    if isinstance(cancel_result, dict) and "clientOrderID" in cancel_result["data"]:
KeyError: 'data'
```

![image](https://github.com/user-attachments/assets/55d89711-a17e-406f-908c-3995cc6fa8b5)

Danilo-Araujo-Silva commented:
@nikspz

Thank you for your detailed feedback on PR7280! Based on your review, we’ve implemented fixes for the issues you identified, including inconsistent order amounts and cancellation errors. 

Fixes Implemented
Inconsistent Order Amounts:
We created the quantize_order_amount method to ensure that order amounts strictly follow the exchange’s step size requirements. Orders now match the amounts configured in the strategy without deviations (e.g., 5 instead of 4.99).

Failed Order Cancellation:
Adjustments to the _place_cancel method addressed issues with cancellation responses. It now properly handles timestamps and integrates smoothly with Hummingbot’s OrderUpdate process.
This ensures that canceled orders are correctly reflected within Hummingbot and the exchange.

Order Visibility on BingX Website:
Previously, some canceled orders didn’t immediately disappear from the BingX website, even though refreshing resolved the issue. After our updates, this behavior no longer seems to occur, and orders disappear as expected after cancellation.

Current Status
Orders now adhere to the exact amounts specified in the strategy.
Cancellation requests are correctly processed during the strategy run.
The "Failed to cancel the following orders" message after stopping the bot should no longer appear, provided the exchange processes cancellations successfully.

We’d appreciate it if you could test the updated connector and share any additional observations or suggestions for improvement. Let us know if anything else requires attention!

Thank you for your support and collaboration in improving this connector.

danilo-robotter-ai commented:
@nikspz

Thank you for your detailed feedback on PR7280! Based on your review, we’ve implemented fixes for the issues you identified, including inconsistent order amounts and cancellation errors. 

Fixes Implemented
Inconsistent Order Amounts:
We created the quantize_order_amount method to ensure that order amounts strictly follow the exchange’s step size requirements. Orders now match the amounts configured in the strategy without deviations (e.g., 5 instead of 4.99).

Failed Order Cancellation:
Adjustments to the _place_cancel method addressed issues with cancellation responses. It now properly handles timestamps and integrates smoothly with Hummingbot’s OrderUpdate process.
This ensures that canceled orders are correctly reflected within Hummingbot and the exchange.

Order Visibility on BingX Website:
Previously, some canceled orders didn’t immediately disappear from the BingX website, even though refreshing resolved the issue. After our updates, this behavior no longer seems to occur, and orders disappear as expected after cancellation.

Current Status
Orders now adhere to the exact amounts specified in the strategy.
Cancellation requests are correctly processed during the strategy run.
The "Failed to cancel the following orders" message after stopping the bot should no longer appear, provided the exchange processes cancellations successfully.

We’d appreciate it if you could test the updated connector and share any additional observations or suggestions for improvement. Let us know if anything else requires attention!

Thank you for your support and collaboration in improving this connector.

nikspz commented:
Issue with orders stuck / lost on exchange was fixed ✅

Current issues:

Steps to reproduce:
1. Clone and install PR7280
2. connect bing_x
3. create/start pureMM strategy with tight spread
4. review logs

Actual:

1. After filled order event retunr an issue - ERROR - Unexpected error while processing event 200.

[logs_bingx-2deb.log](https://github.com/user-attachments/files/18038462/logs_bingx-2deb.log)
[logs_bingxrp.log](https://github.com/user-attachments/files/18038464/logs_bingxrp.log)


```
2024-12-06 12:05:04,675 - 2104918 - hummingbot.connector.client_order_tracker - INFO - The BUY order BINGX_BXPUT62898d142007a909c5d13 amounting to 2.990000000000000213162820728/2.990000000000000062241878318 XRP has been filled at 2.2980999999999998095745468162931501865386962890625 USDT.
2024-12-06 12:05:04,676 - 2104918 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (XRP-USDT) Maker buy order of 2.9900000000000002131628207280300557613372802734375 XRP filled. [clock=2024-12-06 12:05:04+00:00]
2024-12-06 12:05:04,686 - 2104918 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1733486704.0, "order_id": "BINGX_BXPUT62898d142007a909c5d13", "trading_pair": "XRP-USDT", "trade_type": "TradeType.BUY", "order_type": "OrderType.LIMIT", "price": "2.2980999999999998095745468162931501865386962890625", "amount": "2.9900000000000002131628207280300557613372802734375", "trade_fee": {"percent": "0", "percent_token": null, "flat_fees": [{"token": "XRP", "amount": "-0.002990000000000000469901895172597505734302103519439697265625"}]}, "exchange_trade_id": "108025595", "exchange_order_id": "1865004390150995968", "leverage": 1, "position": "NIL", "event_name": "OrderFilledEvent", "event_source": "bing_x"}
2024-12-06 12:05:04,687 - 2104918 - hummingbot.connector.client_order_tracker - INFO - Created LIMIT BUY order BINGX_BXPUT62898d142007a909c5d13 for 2.990000000000000062241878318 XRP-USDT at 2.298100000000000110128941556.
2024-12-06 12:05:04,691 - 2104918 - hummingbot.connector.exchange.bing_x.bing_x_exchange.BingXExchange - ERROR - Unexpected error while processing event 200.
Traceback (most recent call last):
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1910, in _execute_context
    self.dialect.do_execute(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
sqlite3.IntegrityError: UNIQUE constraint failed: Order.id

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "hummingbot/core/pubsub.pyx", line 165, in hummingbot.core.pubsub.PubSub.c_trigger_event
    typed_listener.c_call(arg)
  File "hummingbot/core/event/event_listener.pyx", line 25, in hummingbot.core.event.event_listener.EventListener.c_call
    self(arg)
  File "/home/hbot02/nikita/bingx2/hummingbot/core/event/event_forwarder.py", line 24, in __call__
    self._to_function(self.current_event_tag, self.current_event_caller, arg)
  File "/home/hbot02/nikita/bingx2/hummingbot/connector/markets_recorder.py", line 332, in _did_create_order
    self.save_market_states(self._config_file_path, market, session=session)
  File "/home/hbot02/nikita/bingx2/hummingbot/connector/markets_recorder.py", line 265, in save_market_states
    market_states: Optional[MarketState] = self.get_market_states(config_file_path, market, session=session)
  File "/home/hbot02/nikita/bingx2/hummingbot/connector/markets_recorder.py", line 293, in get_market_states
    market_states: Optional[MarketState] = query.one_or_none()
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2850, in one_or_none
    return self._iter().one_or_none()
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2916, in _iter
    result = self.session.execute(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 1665, in execute
    ) = compile_state_cls.orm_pre_session_exec(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/context.py", line 312, in orm_pre_session_exec
    session._autoflush()
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2264, in _autoflush
    util.raise_(e, with_traceback=sys.exc_info()[2])
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2253, in _autoflush
    self.flush()
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 3449, in flush
    self._flush(objects)
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 3588, in _flush
    with util.safe_reraise():
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 3549, in _flush
    flush_context.execute()
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/unitofwork.py", line 456, in execute
    rec.execute(self)
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/unitofwork.py", line 630, in execute
    util.preloaded.orm_persistence.save_obj(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/persistence.py", line 245, in save_obj
    _emit_insert_statements(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/persistence.py", line 1097, in _emit_insert_statements
    c = connection._execute_20(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1710, in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/sql/elements.py", line 334, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1577, in _execute_clauseelement
    ret = self._execute_context(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1953, in _execute_context
    self._handle_dbapi_exception(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 2134, in _handle_dbapi_exception
    util.raise_(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1910, in _execute_context
    self.dialect.do_execute(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.IntegrityError: (raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely)
(sqlite3.IntegrityError) UNIQUE constraint failed: Order.id
[SQL: INSERT INTO "Order" (id, config_file_path, strategy, market, symbol, base_asset, quote_asset, creation_timestamp, order_type, amount, leverage, price, last_status, last_update_timestamp, exchange_order_id, position) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)]
[parameters: ('BINGX_BXPUT62898d142007a909c5d13', 'bingx-2deb.yml', 'pure_market_making', 'bing_x', 'XRP-USDT', 'XRP', 'USDT', 1733486671000, 'LIMIT', 2990000, 1, 2298100, 'BuyOrderCreated', 1733486671000, '1865004390150995968', 'NIL')]
(Background on this error at: https://sqlalche.me/e/14/gkpj)
```


2. After Cancelling Limit order event have an issue AttributeError: 'str' object has no attribute 'value'

[logs_bingx-2deb.log](https://github.com/user-attachments/files/18038475/logs_bingx-2deb.log)
[logs_bingxrp.log](https://github.com/user-attachments/files/18038472/logs_bingxrp.log)


```
2024-12-06 12:43:09,674 - 2104918 - hummingbot.core.utils.async_utils - ERROR - Unhandled error in background task: 'str' object has no attribute 'value'
Traceback (most recent call last):
  File "/home/hbot02/nikita/bingx2/hummingbot/core/utils/async_utils.py", line 9, in safe_wrapper
    return await c
  File "/home/hbot02/nikita/bingx2/hummingbot/connector/client_order_tracker.py", line 292, in _process_order_update
    updated: bool = tracked_order.update_with_order_update(order_update)
  File "/home/hbot02/nikita/bingx2/hummingbot/core/data_type/in_flight_order.py", line 330, in update_with_order_update
    self.check_processed_by_exchange_condition()
  File "/home/hbot02/nikita/bingx2/hummingbot/core/data_type/in_flight_order.py", line 369, in check_processed_by_exchange_condition
    if self.current_state.value > OrderState.PENDING_CREATE.value:
AttributeError: 'str' object has no attribute 'value'
```

3. Orders amount is not matched the amounts configured in the strategy (order amount is 3, placing with 2.99, or 0.0499 / 0.05)


ERROR - Unhandled error in background task: 'str' object has no attribute 'value':
![image](https://github.com/user-attachments/assets/bd47de9b-ac73-443f-a9ab-f573e71710dd)
ERROR - Unexpected error while processing event 200.
![image](https://github.com/user-attachments/assets/ffecd355-32c9-4099-a8b5-e24582e9d8af)
ERROR - Unhandled error in background task: 'str' object has no attribute 'value': after stop command
![image](https://github.com/user-attachments/assets/afc54523-8e00-4eef-88ff-48f0d6b6ad83)

Order amount / price not matched 
![image](https://github.com/user-attachments/assets/8d4b97fc-a0a4-430b-a03f-32e9ea8e9f42)
![image](https://github.com/user-attachments/assets/6535d20e-467a-48e8-bc71-772eb90c0027)

danilo-robotter-ai commented:
## Update on BingX Connector Improvements

Hi @nikspz,

Thank you for your continued feedback. We've implemented additional fixes and updates based on the issues you raised in your recent review.

### Addressed Issues

1. **Error: `Unexpected error while processing event 200`**
   - Resolved by ensuring proper handling of order state transitions. Specifically, we updated the `OrderState` to correctly reflect changes (e.g., `CANCELED` or `FILLED`).
   - This prevents errors during event processing and ensures the state is consistent across the connector.

2. **Error: `'str' object has no attribute 'value'`**
   - Fixed by validating the type of the `OrderState` property to ensure it is always an `OrderState` instance and not a raw string.

3. **Order Amount Mismatch**
   - We improved the accuracy of order amounts by updating the `_format_trading_rules` method. The changes ensure that `stepSize`, `tickSize`, and other parameters from the exchange are applied correctly to the trading rules.
   - Then the `quantize_order_amount` method correctly format amounts according to the trading rules.

   ** Additionally when creating orders, the exchange interface takes a few seconds to display the newly created orders. We recommend waiting a few seconds or refreshing the page to see the updates. This issue does not occur with canceled orders, which are reflected immediately. **

Let us know if there are any remaining issues.

![image](https://github.com/user-attachments/assets/22b499f0-cea6-4b74-9fa1-e5396a6bde30)

nikspz commented:
Tests:
- Cloned and installed 7280
- connected bing_x
- checked balance
- Market availability during the strategy creation:  OK
- Compatibility with the available strategies: OK
- status, order_book and status --live updates:  Note: there’s delay updating best bid/ best ask and  total / available balance
- Order submission and cancellation:  OK
- Filled orders: 
  - Review Unexpected error while processing event 200, after buy order filled
  - **ERROR** - Unexpected error while processing event 201, after sell order filled
- history: Review Client recorded one filled order (BUY) event twice


Note: there’s delay updating best bid/ best ask and  total / available balance


[logs_conf_pure_mm_1.log](https://github.com/user-attachments/files/18125601/logs_conf_pure_mm_1.log)
[b3.zip](https://github.com/user-attachments/files/18125637/b3.zip)


```
2024-12-13 10:17:26,001 - 2265199 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (SOL-USDT) Creating 2 bid orders at (Size, Price): ['0.05 SOL, 225.32 USDT', '0.05 SOL, 223.07 USDT']
2024-12-13 10:17:26,322 - 2265199 - hummingbot.connector.client_order_tracker - INFO - Created LIMIT BUY order BINGX_BSLUT62924232da1b295a3a473 for 0.050 SOL-USDT at 225.32.
2024-12-13 10:17:26,340 - 2265199 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1734085046.0, "type": "OrderType.LIMIT", "trading_pair": "SOL-USDT", "amount": "0.050", "price": "225.32", "order_id": "BINGX_BSLUT62924232da1b295a3a473", "creation_timestamp": 1734085046.0, "exchange_order_id": "1867514155867471872", "leverage": 1, "position": "NIL", "event_name": "BuyOrderCreatedEvent", "event_source": "bing_x"}
2024-12-13 10:17:26,450 - 2265199 - hummingbot.connector.client_order_tracker - INFO - Created LIMIT BUY order BINGX_BSLUT62924232da33295a3a473 for 0.050 SOL-USDT at 223.07.
2024-12-13 10:17:26,470 - 2265199 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1734085046.0, "type": "OrderType.LIMIT", "trading_pair": "SOL-USDT", "amount": "0.050", "price": "223.07", "order_id": "BINGX_BSLUT62924232da33295a3a473", "creation_timestamp": 1734085046.0, "exchange_order_id": "1867514156383371264", "leverage": 1, "position": "NIL", "event_name": "BuyOrderCreatedEvent", "event_source": "bing_x"}
2024-12-13 10:18:35,545 - 2265199 - hummingbot.connector.client_order_tracker - INFO - The BUY order BINGX_BSLUT62924232da1b295a3a473 amounting to 0.007000000000000000145716771982/0.050 SOL has been filled at 225.31999999999999317878973670303821563720703125 USDT.
2024-12-13 10:18:35,546 - 2265199 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (SOL-USDT) Maker buy order of 0.007000000000000000145716771982051795930601656436920166015625 SOL filled. [clock=2024-12-13 10:18:35+00:00]
2024-12-13 10:18:35,577 - 2265199 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1734085115.0, "order_id": "BINGX_BSLUT62924232da1b295a3a473", "trading_pair": "SOL-USDT", "trade_type": "TradeType.BUY", "order_type": "OrderType.LIMIT", "price": "225.31999999999999317878973670303821563720703125", "amount": "0.007000000000000000145716771982051795930601656436920166015625", "trade_fee": {"percent": "0", "percent_token": null, "flat_fees": [{"token": "SOL", "amount": "-0.0000069999999999999998949950195947788955663781962357461452484130859375"}]}, "exchange_trade_id": "119851165", "exchange_order_id": "1867514155867471872", "leverage": 1, "position": "NIL", "event_name": "OrderFilledEvent", "event_source": "bing_x"}
2024-12-13 10:18:35,578 - 2265199 - hummingbot.connector.client_order_tracker - INFO - Created LIMIT BUY order BINGX_BSLUT62924232da1b295a3a473 for 0.050 SOL-USDT at 225.32.
2024-12-13 10:18:35,597 - 2265199 - hummingbot.connector.exchange.bing_x.bing_x_exchange.BingXExchange - ERROR - Unexpected error while processing event 200.
Traceback (most recent call last):
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1910, in _execute_context
    self.dialect.do_execute(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
sqlite3.IntegrityError: UNIQUE constraint failed: Order.id

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "hummingbot/core/pubsub.pyx", line 165, in hummingbot.core.pubsub.PubSub.c_trigger_event
    typed_listener.c_call(arg)
  File "hummingbot/core/event/event_listener.pyx", line 25, in hummingbot.core.event.event_listener.EventListener.c_call
    self(arg)
  File "/home/hbot02/nikita/bingx2/hummingbot/core/event/event_forwarder.py", line 24, in __call__
    self._to_function(self.current_event_tag, self.current_event_caller, arg)
  File "/home/hbot02/nikita/bingx2/hummingbot/connector/markets_recorder.py", line 332, in _did_create_order
    self.save_market_states(self._config_file_path, market, session=session)
  File "/home/hbot02/nikita/bingx2/hummingbot/connector/markets_recorder.py", line 265, in save_market_states
    market_states: Optional[MarketState] = self.get_market_states(config_file_path, market, session=session)
  File "/home/hbot02/nikita/bingx2/hummingbot/connector/markets_recorder.py", line 293, in get_market_states
    market_states: Optional[MarketState] = query.one_or_none()
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2850, in one_or_none
    return self._iter().one_or_none()
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2916, in _iter
    result = self.session.execute(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 1665, in execute
    ) = compile_state_cls.orm_pre_session_exec(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/context.py", line 312, in orm_pre_session_exec
    session._autoflush()
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2264, in _autoflush
    util.raise_(e, with_traceback=sys.exc_info()[2])
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2253, in _autoflush
    self.flush()
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 3449, in flush
    self._flush(objects)
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 3588, in _flush
    with util.safe_reraise():
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 3549, in _flush
    flush_context.execute()
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/unitofwork.py", line 456, in execute
    rec.execute(self)
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/unitofwork.py", line 630, in execute
    util.preloaded.orm_persistence.save_obj(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/persistence.py", line 245, in save_obj
    _emit_insert_statements(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/orm/persistence.py", line 1097, in _emit_insert_statements
    c = connection._execute_20(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1710, in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/sql/elements.py", line 334, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1577, in _execute_clauseelement
    ret = self._execute_context(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1953, in _execute_context
    self._handle_dbapi_exception(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 2134, in _handle_dbapi_exception
    util.raise_(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1910, in _execute_context
    self.dialect.do_execute(
  File "/home/hbot02/miniconda3/envs/hummingbot/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.IntegrityError: (raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely)
(sqlite3.IntegrityError) UNIQUE constraint failed: Order.id
[SQL: INSERT INTO "Order" (id, config_file_path, strategy, market, symbol, base_asset, quote_asset, creation_timestamp, order_type, amount, leverage, price, last_status, last_update_timestamp, exchange_order_id, position) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)]
[parameters: ('BINGX_BSLUT62924232da1b295a3a473', 'conf_pure_mm_1.yml', 'pure_market_making', 'bing_x', 'SOL-USDT', 'SOL', 'USDT', 1734085046000, 'LIMIT', 50000, 1, 225320000, 'BuyOrderCreated', 1734085046000, '1867514155867471872', 'NIL')]
(Background on this error at: https://sqlalche.me/e/14/gkpj)
2024-12-13 10:18:35,672 - 2265199 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1734085115.0, "type": "OrderType.LIMIT", "trading_pair": "SOL-USDT", "amount": "0.050", "price": "225.32", "order_id": "BINGX_BSLUT62924232da1b295a3a473", "creation_timestamp": 1734085046.0, "exchange_order_id": "1867514155867471872", "leverage": 1, "position": "NIL", "event_name": "BuyOrderCreatedEvent", "event_source": "bing_x"}
2024-12-13 10:18:56,045 - 2265199 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (SOL-USDT) Canceling the limit order BINGX_BSLUT62924232da1b295a3a473. [clock=2024-12-13 10:18:56+00:00]
2024-12-13 10:18:56,045 - 2265199 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (SOL-USDT) Canceling the limit order BINGX_BSLUT62924232da33295a3a473. [clock=2024-12-13 10:18:56+00:00]
2024-12-13 10:18:56,268 - 2265199 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1734085136.0, "order_id": "BINGX_BSLUT62924232da1b295a3a473", "exchange_order_id": "1867514155867471872", "event_name": "OrderCancelledEvent", "event_source": "bing_x"}
2024-12-13 10:18:56,269 - 2265199 - hummingbot.connector.client_order_tracker - INFO - Successfully canceled order BINGX_BSLUT62924232da1b295a3a473.
2024-12-13 10:18:56,387 - 2265199 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1734085136.0, "order_id": "BINGX_BSLUT62924232da33295a3a473", "exchange_order_id": "1867514156383371264", "event_name": "OrderCancelledEvent", "event_source": "bing_x"}
2024-12-13 10:18:56,387 - 2265199 - hummingbot.connector.client_order_tracker - INFO - Successfully canceled order BINGX_BSLUT62924232da33295a3a473.
2024-12-13 10:18:57,046 - 2265199 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (SOL-USDT) Creating 2 bid orders at (Size, Price): ['0.05 SOL, 225.09 USDT', '0.05 SOL, 222.84 USDT']
2024-12-13 10:18:57,047 - 2265199 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (SOL-USDT) Creating 1 ask orders at (Size, Price): ['0.006 SOL, 225.66 USDT']
```


![image](https://github.com/user-attachments/assets/7d16c02e-b128-434d-8bbe-78809bc86d01)

![image](https://github.com/user-attachments/assets/397815c8-3356-4c1a-a5ae-c43d219ef831)

nikspz commented:
Let me check if same thing on 0fa6ced5ab03d6e5547ce905946b09ff8b128b1b

nikspz commented:
commit 0fa6ced

Got same issues as here https://github.com/hummingbot/hummingbot/pull/7280#issuecomment-2541232608

1. Unexpected error while processing event 200, after buy order filled
ERROR - Unexpected error while processing event 201, after sell order filled

[logs_conf_pure_mm_1.log](https://github.com/user-attachments/files/18131906/logs_conf_pure_mm_1.log)
[b4.zip](https://github.com/user-attachments/files/18131914/b4.zip)



2. Duplicated orders in history:
![image](https://github.com/user-attachments/assets/83e51443-5e0d-40f7-9247-50eb479637be)
![image](https://github.com/user-attachments/assets/518afbac-54ae-4d2b-8580-746383c6e646)

Note: there’s delay updating best bid/ best ask and total / available balance

![bingx-sol-orderbook](https://github.com/user-attachments/assets/66e74865-b75e-4126-8ce7-e4d197f6a360)

danilo-robotter-ai commented:
Hi @nikspz,

Thanks once again for your valuable feedback!

The previous errors have been more difficult to track since they do not always occur (where they seem to arise when events are processed in parallel). 

Because of the nature of the error, what we did with this new approach was to create additional safeguards in the code to try to prevent some events from being processed before others.

For example, the main error (`sqlite3.IntegrityError: UNIQUE constraint failed: Order.id`) seems to happen when a `filled` event is processed before an `open` one, but both are being processed altogether. We had left the strategy to run for the entire night and the error seems to have disappeared.

About the order book live, we have done some study, and seems better now, sometimes a little delay, but usually it updates quite quickly, please kindly let us know if you find inconsistencies in this regard.

Finally, we have noticed some errors related to WebSocket disconnection or miscommunication, but they are probably inevitable and the system usually can recover itself from them.

If you happen to find other inconsistencies, please let us know.

Thanks a lot in advance!

nikspz reviewed (approved):
Tests:

- Cloned and installed 7280
- connected bing_x
- checked balance
- Market availability during the strategy creation: OK
- Compatibility with the available strategies: OK
- created/ started successfully pureMM and pmm_simple v2
- status, order_book and status --live updates: OK
- Order submission and cancellation: OK
- Fees matched with exchange
- Filled orders: Limit/ market OK

cardosofede reviewed (approved):
LGTM but there is a lot of commented code that should be removed. Also I will recommend to implement error handling for time sync issues
